# ❤️ Heart Disease Prediction System using Machine Learning




---

# 📄 Project Metadata
### **Title:** **Heart Disease Prediction using Machine Learning** ❤️‍🩹
### **Author:** **Asad Ali** ✍️
### **Institute:** **University of Okara & Trainee at Codanics** 🎓
### **Email:** 📧 [asadalyy834@gmail.com](mailto:asadalyy834@gmail.com)
### **Course:** **Data Science / Machine Learning** 📊🧠
### **Date:** **July 2025** 📅
### **Version:** **1.0** 🔢
### **Language:** **Python** 🐍
### **Libraries Overview:**
- **Pandas** 📚: Data manipulation and analysis
- **NumPy** 🔢: Numerical computing
- **Scikit-learn** ⚙️: Machine learning tools and algorithms
- **Matplotlib** 📈: Data visualization
- **Seaborn** 🌊: Statistical data visualization
- **Plotly** 📊: Interactive data visualization
### **Dataset:** **UCI Heart Disease Dataset from Kaggle** 📋

---


# 📊 About the Dataset

### 🧬 Context

This is a **multivariate dataset** — it contains multiple statistical variables and supports numerical data analysis. Although the full dataset includes **76 attributes**, most published research focuses on **14 key features**.

- 📍 **Primary Source Used:** *Cleveland database* — the most commonly used by machine learning researchers.
- 🧠 **Main Objective:** Predict whether a patient has heart disease or not based on medical parameters.
- 🔎 **Secondary Objective:** Gain diagnostic insights through statistical and machine learning exploration.

---

### 📌 Selected Attribute Descriptions (14 Core Features)

| 🔢 No. | 🧬 Column Name | 📖 Description                                                                 |
|-------:|---------------|--------------------------------------------------------------------------------|
| 1️⃣    | `age`         | Age of the patient (in years)                                                  |
| 2️⃣    | `sex`         | Gender of patient (`0` = Female, `1` = Male)                                   |
| 3️⃣    | `cp`          | Chest pain type: `typical angina`, `atypical angina`, `non-anginal`, `asymptomatic` |
| 4️⃣    | `trestbps`    | Resting blood pressure (in mm Hg at admission)                                 |
| 5️⃣    | `chol`        | Serum cholesterol level (in mg/dl)                                             |
| 6️⃣    | `fbs`         | Fasting blood sugar > 120 mg/dl (`1` = True; `0` = False)                       |
| 7️⃣    | `restecg`     | ECG results: `normal`, `ST-T abnormality`, `left ventricular hypertrophy`      |
| 8️⃣    | `thalach`     | Maximum heart rate achieved                                                    |
| 9️⃣    | `exang`       | Exercise-induced angina (`1` = Yes; `0` = No)                                   |
| 🔟     | `oldpeak`     | ST depression induced by exercise relative to rest                             |
| 1️⃣1️⃣ | `slope`       | Slope of the peak exercise ST segment                                          |
| 1️⃣2️⃣ | `ca`          | Number of major vessels (0–3) colored by fluoroscopy                           |
| 1️⃣3️⃣ | `thal`        | Thalassemia condition: `normal`, `fixed defect`, `reversible defect`           |
| 1️⃣4️⃣ | `target/num`  | Predicted attribute (0 = No Disease, 1 = Heart Disease)                         |

---

### 🧾 Additional Columns (May Appear in Extended Datasets)

| 🔹 Column        | 🔍 Description                              |
|------------------|--------------------------------------------|
| `id`             | Unique ID for each patient                 |
| `dataset`         | Source location of data (e.g., Hungary)   |

---

### 👨‍⚕️ Acknowledgements

**Contributors & Medical Institutions:**

- 🏥 *Hungarian Institute of Cardiology, Budapest*: **Dr. Andras Janosi**  
- 🏥 *University Hospital, Zurich, Switzerland*: **Dr. William Steinbrunn**  
- 🏥 *University Hospital, Basel, Switzerland*: **Dr. Matthias Pfisterer**  
- 🏥 *V.A. Medical Center, Long Beach & Cleveland Clinic*: **Dr. Robert Detrano**

---

### 📚 Relevant Research Papers

- 📄 *International application of a new probability algorithm for the diagnosis of coronary artery disease*  
  ➤ *Detrano, R. et al., American Journal of Cardiology, 1989*

- 📄 *Instance-based prediction of heart-disease presence with the Cleveland database*  
  ➤ *David W. Aha & Dennis Kibler*

- 📄 *Models of incremental concept formation*  
  ➤ *Gennari, J.H., Langley, P., & Fisher, D., Artificial Intelligence, 1989*

---

### 🙏 Citation Request

> The authors request that any publication using this dataset must credit the principal investigators:
> 
> - **Dr. Andras Janosi** – Hungarian Institute of Cardiology  
> - **Dr. William Steinbrunn** – University Hospital, Zurich  
> - **Dr. Matthias Pfisterer** – University Hospital, Basel  
> - **Dr. Robert Detrano** – Cleveland Clinic Foundation & Long Beach VA Medical Center

---



# 1. 📚 Importing Libraries

In [373]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.impute import KNNImputer
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [374]:
# Setting to Display max rows and columns
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
# Setting to ignore warnings
import warnings
warnings.filterwarnings("ignore")

In [375]:
# Let's Load the Dataset that is in our local directory
df = pd.read_csv("heart_disease_uci.csv")
# Let's have a look at the first few rows of the dataset
print(df.head())

   id  age     sex    dataset               cp  trestbps   chol    fbs  \
0   1   63    Male  Cleveland   typical angina     145.0  233.0   True   
1   2   67    Male  Cleveland     asymptomatic     160.0  286.0  False   
2   3   67    Male  Cleveland     asymptomatic     120.0  229.0  False   
3   4   37    Male  Cleveland      non-anginal     130.0  250.0  False   
4   5   41  Female  Cleveland  atypical angina     130.0  204.0  False   

          restecg  thalch  exang  oldpeak        slope   ca  \
0  lv hypertrophy   150.0  False      2.3  downsloping  0.0   
1  lv hypertrophy   108.0   True      1.5         flat  3.0   
2  lv hypertrophy   129.0   True      2.6         flat  2.0   
3          normal   187.0  False      3.5  downsloping  0.0   
4  lv hypertrophy   172.0  False      1.4    upsloping  0.0   

                thal  num  
0       fixed defect    0  
1             normal    2  
2  reversable defect    1  
3             normal    0  
4             normal    0  


In [376]:
# Getting the info of our Dataset
print("Information about the Dataset")
print(df.info())

Information about the Dataset
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 920 entries, 0 to 919
Data columns (total 16 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   id        920 non-null    int64  
 1   age       920 non-null    int64  
 2   sex       920 non-null    object 
 3   dataset   920 non-null    object 
 4   cp        920 non-null    object 
 5   trestbps  861 non-null    float64
 6   chol      890 non-null    float64
 7   fbs       830 non-null    object 
 8   restecg   918 non-null    object 
 9   thalch    865 non-null    float64
 10  exang     865 non-null    object 
 11  oldpeak   858 non-null    float64
 12  slope     611 non-null    object 
 13  ca        309 non-null    float64
 14  thal      434 non-null    object 
 15  num       920 non-null    int64  
dtypes: float64(5), int64(3), object(8)
memory usage: 115.1+ KB
None


In [377]:
# Check the Shape of the Dataset
print("Shape of the Dataset")
print("The Dataset has", df.shape[0], "rows and", df.shape[1], "columns.")

Shape of the Dataset
The Dataset has 920 rows and 16 columns.


-----

# 2. Data Preprocessing 🔍
## 1. **Handling Missing Values:** Fill or drop missing data.




In [378]:
# Let's First check for missing values in our dataset
missing_values = df.isnull().sum().sort_values(ascending=False)
print("Missing Values in the Dataset:")
print(missing_values[missing_values > 0])

Missing Values in the Dataset:
ca          611
thal        486
slope       309
fbs          90
oldpeak      62
trestbps     59
exang        55
thalch       55
chol         30
restecg       2
dtype: int64


In [379]:
# Check the percentage of missing values in our dataset
print((df.isnull().sum() / len(df) * 100).sort_values(ascending=False))

ca          66.413043
thal        52.826087
slope       33.586957
fbs          9.782609
oldpeak      6.739130
trestbps     6.413043
exang        5.978261
thalch       5.978261
chol         3.260870
restecg      0.217391
cp           0.000000
dataset      0.000000
id           0.000000
age          0.000000
sex          0.000000
num          0.000000
dtype: float64


In [380]:
# Let's check the type of each column that has missing values in our dataset.
print("Data Types of Columns with Missing Values:")
print(df.dtypes[missing_values[missing_values > 0].index])

Data Types of Columns with Missing Values:
ca          float64
thal         object
slope        object
fbs          object
oldpeak     float64
trestbps    float64
exang        object
thalch      float64
chol        float64
restecg      object
dtype: object


In [381]:
# Let's impute missing values that have int or float data types using KNN Imputer
imputer = KNNImputer(n_neighbors=5)
df_imputed = pd.DataFrame(imputer.fit_transform(df.select_dtypes(include=[np.number])), columns=df.select_dtypes(include=[np.number]).columns)
# Replacing the original numeric columns with the imputed ones
df[df.select_dtypes(include=[np.number]).columns] = df_imputed
# Let's again have a look at the missing values after imputation of numeric columns
missing_values_after_imputation = df.isnull().sum().sort_values(ascending=False)
print("Missing Values in the Dataset after Imputation:")
print(missing_values_after_imputation[missing_values_after_imputation > 0])


Missing Values in the Dataset after Imputation:
thal       486
slope      309
fbs         90
exang       55
restecg      2
dtype: int64


In [382]:
# Let's again check the datatype of each column that contain missing values after imputation of numeric columns
print("Data Types of Columns with Missing Values after Imputation:")
print(df.dtypes[missing_values_after_imputation[missing_values_after_imputation > 0].index])


Data Types of Columns with Missing Values after Imputation:
thal       object
slope      object
fbs        object
exang      object
restecg    object
dtype: object


In [383]:
# Encode , Impute and Decode using ML Model
def encode_and_rf_impute(df):
    df_copy = df.copy()
    # Encode object columns
    label_encoders = {}
    object_cols = df_copy.select_dtypes(include=['object']).columns
    for col in object_cols:
        le = LabelEncoder()
        df_copy[col] = le.fit_transform(df_copy[col].astype(str))
        label_encoders[col] = le

    # Impute object columns using RandomForestClassifier
    from sklearn.ensemble import RandomForestClassifier
    for col in object_cols:
        missing = df[col].isnull()
        if missing.any():
            not_missing = ~missing
            X_train = df_copy.loc[not_missing].drop(columns=[col])
            y_train = df_copy.loc[not_missing, col]
            X_pred = df_copy.loc[missing].drop(columns=[col])
            clf = RandomForestClassifier(n_estimators=100, random_state=0)
            clf.fit(X_train, y_train)
            df_copy.loc[missing, col] = clf.predict(X_pred)

    # Decode object columns back to original values
    for col, le in label_encoders.items():
        df_copy[col] = le.inverse_transform(df_copy[col].round().astype(int))

    return df_copy

# Apply the function using RandomForestClassifier for categorical columns
df  = encode_and_rf_impute(df)
print("Missing Values in the Dataset after Encoding and RF Imputation:")
print(df.isnull().sum())


Missing Values in the Dataset after Encoding and RF Imputation:
id          0
age         0
sex         0
dataset     0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalch      0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
num         0
dtype: int64


In [384]:
# Let's again check the few random rows of the dataset after encoding and imputation
print("Random Rows of the Dataset after Encoding and Imputation:")
print(df.sample(5))

Random Rows of the Dataset after Encoding and Imputation:
        id   age     sex        dataset               cp  trestbps   chol  \
839  840.0  56.0    Male  VA Long Beach     asymptomatic     116.4    0.0   
455  456.0  54.0    Male        Hungary  atypical angina     160.0  305.0   
908  909.0  74.0    Male  VA Long Beach     asymptomatic     155.0  310.0   
718  719.0  73.0  Female    Switzerland      non-anginal     160.0    0.0   
832  833.0  49.0    Male  VA Long Beach      non-anginal     129.8  142.0   

       fbs           restecg  thalch  exang  oldpeak        slope   ca  \
839  False    lv hypertrophy   125.4   True      1.8  downsloping  1.0   
455  False            normal   175.0  False      0.0    upsloping  0.2   
908  False            normal   112.0   True      1.5  downsloping  1.0   
718  False  st-t abnormality   121.0  False      0.0    upsloping  1.4   
832  False            normal   129.8   True      1.2         flat  1.0   

                  thal  num  
839 

In [385]:
# Getting the info of our dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 920 entries, 0 to 919
Data columns (total 16 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   id        920 non-null    float64
 1   age       920 non-null    float64
 2   sex       920 non-null    object 
 3   dataset   920 non-null    object 
 4   cp        920 non-null    object 
 5   trestbps  920 non-null    float64
 6   chol      920 non-null    float64
 7   fbs       920 non-null    object 
 8   restecg   920 non-null    object 
 9   thalch    920 non-null    float64
 10  exang     920 non-null    object 
 11  oldpeak   920 non-null    float64
 12  slope     920 non-null    object 
 13  ca        920 non-null    float64
 14  thal      920 non-null    object 
 15  num       920 non-null    float64
dtypes: float64(8), object(8)
memory usage: 115.1+ KB


- ###  Missing values ka rola howa khtm 😎 

## 2. **Cleaning:** Remove duplicates and handle outliers.

In [386]:
# Let's check for duplicates in our dataset
duplicates = df.duplicated().sum()
print(f"Number of duplicate rows in the dataset: {duplicates}")

Number of duplicate rows in the dataset: 0


- ### There is no Duplicates in our Dataset.

**🏁 Data Cleaning Mission: Complete!**

All data cleaning steps have been successfully performed:
- Duplicate records have been checked and not found.
- We will deal Outliers in the Exploratory Data Analysis.

Our dataset is now clean, reliable, and ready for Machine learning and Statistical analysis.

# 3. 📊 Exploratory Data Analysis (EDA) 🔍

In [387]:
# Let's look at the tail of Dataset
print(df.tail())

        id   age     sex        dataset               cp  trestbps   chol  \
915  916.0  54.0  Female  VA Long Beach     asymptomatic     127.0  333.0   
916  917.0  62.0    Male  VA Long Beach   typical angina     148.8  139.0   
917  918.0  55.0    Male  VA Long Beach     asymptomatic     122.0  223.0   
918  919.0  58.0    Male  VA Long Beach     asymptomatic     136.2  385.0   
919  920.0  62.0    Male  VA Long Beach  atypical angina     120.0  254.0   

       fbs           restecg  thalch  exang  oldpeak slope   ca  \
915   True  st-t abnormality   154.0  False      0.0  flat  1.0   
916  False  st-t abnormality   133.2   True      1.8  flat  1.0   
917   True  st-t abnormality   100.0  False      0.0  flat  1.2   
918   True    lv hypertrophy   133.4  False      0.6  flat  0.8   
919  False    lv hypertrophy    93.0   True      0.0  flat  1.2   

                  thal  num  
915             normal  1.0  
916  reversable defect  0.0  
917       fixed defect  2.0  
918  reversabl

In [388]:
# Let's investigate the Age column
print("Minimum Age of Patient is", df["age"].min())
print("Minimum Age of Patient is ", df["age"].max())

Minimum Age of Patient is 28.0
Minimum Age of Patient is  77.0


In [389]:
# Taking the number of unique values of Age column
print(df["age"].nunique())

50


In [390]:
# Getting the statistical summary of age.
print(df["age"].describe())

count    920.000000
mean      53.510870
std        9.424685
min       28.000000
25%       47.000000
50%       54.000000
75%       60.000000
max       77.000000
Name: age, dtype: float64


In [391]:
# Let's check the how many males and females have heart disease.
print(df["sex"].value_counts())

sex
Male      726
Female    194
Name: count, dtype: int64


In [392]:
# Total number of males and females in the dataset
sex_counts = df['sex'].value_counts()
male_total = sex_counts.get(0, 0)
female_total = sex_counts.get(1, 0)

print(f"Total males: {male_total}")
print(f"Total females: {female_total}")

# Calculate percentage of males and females
total_people = male_total + female_total
if total_people > 0:
    male_percent = (male_total / total_people) * 100
    female_percent = (female_total / total_people) * 100
    print(f"Percentage of males: {male_percent:.2f}%")
    print(f"Percentage of females: {female_percent:.2f}%")
else:
    print("No Data Available")

Total males: 726
Total females: 194
Percentage of males: 78.91%
Percentage of females: 21.09%


In [393]:
# Let's jump into dataset column
print(df["dataset"].unique())  # This line will show the unique values in the 'dataset' column
# Getting the value count of the 'dataset' column
print(df["dataset"].value_counts())

['Cleveland' 'Hungary' 'Switzerland' 'VA Long Beach']
dataset
Cleveland        304
Hungary          293
VA Long Beach    200
Switzerland      123
Name: count, dtype: int64


In [394]:
# Age bins for 28-77 range
age_bins = [28, 35, 40, 45, 50, 55, 60, 65, 70, 75, 77]
df['age_group'] = pd.cut(df['age'], bins=age_bins, labels=[
    '28-34', '35-39', '40-44', '45-49', '50-54', '55-59', '60-64', '65-69', '70-75', '76-77'
])

# Group by dataset, sex, and age_group for counts
grouped = df.groupby(['dataset', 'sex', 'age_group']).size().reset_index(name='count')

# Plotly bar plot: hover shows gender, age bin, and count
fig = px.bar(
    grouped,
    x='dataset',
    y='count',
    color='sex',
    barmode='group',
    title='Patients Count by Dataset, Gender, and Age Group',
    hover_data=['sex', 'age_group', 'count']
)
fig.update_layout(
    xaxis_title="Dataset",
    yaxis_title="Count",
    legend_title_text="Gender"
)
fig.show()

In [395]:
# Let's again check the first few rows of dataset
print(df.head())


    id   age     sex    dataset               cp  trestbps   chol    fbs  \
0  1.0  63.0    Male  Cleveland   typical angina     145.0  233.0   True   
1  2.0  67.0    Male  Cleveland     asymptomatic     160.0  286.0  False   
2  3.0  67.0    Male  Cleveland     asymptomatic     120.0  229.0  False   
3  4.0  37.0    Male  Cleveland      non-anginal     130.0  250.0  False   
4  5.0  41.0  Female  Cleveland  atypical angina     130.0  204.0  False   

          restecg  thalch  exang  oldpeak        slope   ca  \
0  lv hypertrophy   150.0  False      2.3  downsloping  0.0   
1  lv hypertrophy   108.0   True      1.5         flat  3.0   
2  lv hypertrophy   129.0   True      2.6         flat  2.0   
3          normal   187.0  False      3.5  downsloping  0.0   
4  lv hypertrophy   172.0  False      1.4    upsloping  0.0   

                thal  num age_group  
0       fixed defect  0.0     60-64  
1             normal  2.0     65-69  
2  reversable defect  1.0     65-69  
3           

In [396]:
# Getting the value count of cp
print(df["cp"].value_counts())

cp
asymptomatic       496
non-anginal        204
atypical angina    174
typical angina      46
Name: count, dtype: int64


In [397]:
# This code calculates the median age for each chest pain type and sex, 
# then creates a grouped bar plot to visualize the average age by chest pain type and sex.
# Calculate the average age for each chest pain type and sex
average_age = df.groupby(['cp', 'sex'])['age'].median().round(0).astype(int).reset_index()

# Bar Plot using Plotly
fig = px.bar(average_age, 
             x='cp', 
             y='age', 
             color='sex',
             barmode='group',
             title='Average Age by Chest Pain Type and Sex',
             labels={'age': 'Average Age', 'chest_pain': 'Chest Pain Type'})

# Show the plot
fig.show()

In [398]:
# Let's view the tresbps using boxplot
# Create a box plot
fig = px.box(df, y='trestbps', title='Box Plot of Resting Blood Pressure (trestbps)',
             labels={'trestbps': 'Resting Blood Pressure (trestbps)'})

# Show the plot
fig.show()

# Handling the Outliers of trestbps(Resting Blood Pressure of Patients)                

In [399]:
# I see and study thorougly i decided to drop trestbps column at these specific rows where i found outliers.
df.drop(df[df["trestbps"] == 0].index, inplace=True)
df.drop(df[df["trestbps"] == 80].index, inplace=True)
df.drop(df[(df["trestbps"] == 178) & (df["chol"] == 0)].index, inplace=True)
df.drop(df[(df["trestbps"] == 180) & (df["chol"] == 0)].index, inplace=True)
df.drop(df[df["trestbps"] == 185].index, inplace=True)
df.drop(df[(df["trestbps"] == 200) & (df["chol"] == 0)].index, inplace=True)

In [400]:
# Let's view the tresbps using boxplot
# Create a box plot
fig = px.box(df, y='trestbps', title='Box Plot of Resting Blood Pressure (trestbps) after removing of Outliers',
             labels={'trestbps': 'Resting Blood Pressure (trestbps)'})

# Show the plot
fig.show()

In [401]:
# Let's investigate the trestbps column
print("The average resting blood pressure of patients is", df['trestbps'].mean().astype(int))
print("The maximum resting blood pressure of a patient is", df['trestbps'].max().astype(int))
print("The minimum resting blood pressure of a patient is", df['trestbps'].min().astype(int))

The average resting blood pressure of patients is 132
The maximum resting blood pressure of a patient is 200
The minimum resting blood pressure of a patient is 92


In [402]:
# Calculate the mean resting blood pressure for males
mean_trestbps_male = df[df['sex'] == 'Male']['trestbps'].mean().astype(int)
# Calculate maximum and minimum resting blood pressure for males
max_trestbps_male = df[df['sex'] == 'Male']['trestbps'].max().astype(int)
min_trestbps_male = df[df['sex'] == 'Male']['trestbps'].min().astype(int)

# Calculate the mean resting blood pressure for females
mean_trestbps_female = df[df['sex'] == 'Female']['trestbps'].mean().astype(int)
# Calculate maximum and minimum resting blood pressure for females
max_trestbps_female = df[df['sex'] == 'Female']['trestbps'].max().astype(int)
min_trestbps_female = df[df['sex'] == 'Female']['trestbps'].min().astype(int)

# Display the results
print(f"Mean Resting Blood Pressure for Males: {mean_trestbps_male}")
print(f"Maximum Resting Blood Pressure for Males: {max_trestbps_male}")
print(f"Minimum Resting Blood Pressure for Males: {min_trestbps_male}")
print("-----------------------------")

# Display the results
print(f"Mean Resting Blood Pressure for Females: {mean_trestbps_female}")
print(f"Maximum Resting Blood Pressure for Females: {max_trestbps_female}")
print(f"Minimum Resting Blood Pressure for Females: {min_trestbps_female}")



Mean Resting Blood Pressure for Males: 132
Maximum Resting Blood Pressure for Males: 200
Minimum Resting Blood Pressure for Males: 92
-----------------------------
Mean Resting Blood Pressure for Females: 131
Maximum Resting Blood Pressure for Females: 200
Minimum Resting Blood Pressure for Females: 94


In [403]:
print(df['dataset'].value_counts())

dataset
Cleveland        304
Hungary          293
VA Long Beach    197
Switzerland      118
Name: count, dtype: int64


In [404]:
# Calculating the average resting blood pressure of patients by dataset. 
mean_trestbps_clev = df[df['dataset'] == 'Cleveland']['trestbps'].mean().astype(int)
mean_trestbps_hung = df[df['dataset'] == 'Hungary']['trestbps'].mean().astype(int)
mean_trestbps_va = df[df['dataset'] == 'VA Long Beach']['trestbps'].mean().astype(int)
mean_trestbps_swiss = df[df['dataset'] == 'Switzerland']['trestbps'].mean().astype(int)
print(f"Mean Resting Blood Pressure of Patients of Cleveland: {mean_trestbps_clev}")
print(f"Mean Resting Blood Pressure of Patients of Hungary: {mean_trestbps_hung}")
print(f"Mean Resting Blood Pressure of Patients of VA Long Beech: {mean_trestbps_va}")
print(f"Mean Resting Blood Pressure of Patients of Switzerland: {mean_trestbps_swiss}")
  

Mean Resting Blood Pressure of Patients of Cleveland: 131
Mean Resting Blood Pressure of Patients of Hungary: 132
Mean Resting Blood Pressure of Patients of VA Long Beech: 134
Mean Resting Blood Pressure of Patients of Switzerland: 128


In [405]:
# Prepare data for the pie chart
mean_values = {
    'Cleveland': mean_trestbps_clev,
    'Hungary': mean_trestbps_hung,
    'VA Long Beach': mean_trestbps_va,
    'Switzerland': mean_trestbps_swiss
}

# Create a DataFrame for Plotly
mean_df = pd.DataFrame(mean_values.items(), columns=['Dataset', 'Mean Trestbps'])

# Create a pie chart
fig = px.pie(mean_df, values='Mean Trestbps', names='Dataset', 
             title='Mean Resting Blood Pressure by Dataset')

# Show the plot
fig.show()

In [406]:
# Again look on the Data
print(df.head())

    id   age     sex    dataset               cp  trestbps   chol    fbs  \
0  1.0  63.0    Male  Cleveland   typical angina     145.0  233.0   True   
1  2.0  67.0    Male  Cleveland     asymptomatic     160.0  286.0  False   
2  3.0  67.0    Male  Cleveland     asymptomatic     120.0  229.0  False   
3  4.0  37.0    Male  Cleveland      non-anginal     130.0  250.0  False   
4  5.0  41.0  Female  Cleveland  atypical angina     130.0  204.0  False   

          restecg  thalch  exang  oldpeak        slope   ca  \
0  lv hypertrophy   150.0  False      2.3  downsloping  0.0   
1  lv hypertrophy   108.0   True      1.5         flat  3.0   
2  lv hypertrophy   129.0   True      2.6         flat  2.0   
3          normal   187.0  False      3.5  downsloping  0.0   
4  lv hypertrophy   172.0  False      1.4    upsloping  0.0   

                thal  num age_group  
0       fixed defect  0.0     60-64  
1             normal  2.0     65-69  
2  reversable defect  1.0     65-69  
3           

In [407]:
# Let's jump into the chol column.
print("The average cholestrol level of patients is", df['chol'].mean().astype(int))
print("The maximum cholestrol level of a patient is", df['chol'].max().astype(int))
print("The minimum cholestrol level  of a patient is", df['chol'].min().astype(int))

The average cholestrol level of patients is 202
The maximum cholestrol level of a patient is 603
The minimum cholestrol level  of a patient is 0


In [408]:
# Let's view the cholestrol using interactive boxplot to find outliers
# Create a box plot
fig = px.box(df, y='chol', title='Box Plot of Cholestrol level (chol) of patients',
             labels={'chol': 'Cholestrol Level (chol)'})

# Show the plot
fig.show()

In [410]:
# Dropping some rows from dataset based on my cholestrol study
df.drop(df[df["chol"] == 0].index, inplace=True)
df.drop(df[df["chol"] == 518].index, inplace=True)
df.drop(df[df["chol"] == 529].index, inplace=True)
df.drop(df[df["chol"] == 564].index, inplace=True)
df.drop(df[df["chol"] == 603].index, inplace=True)

In [412]:
# Now i will use IQR method to clean the any other outlier.
Q1 = df['chol'].quantile(0.25)
Q3 = df['chol'].quantile(0.75)
IQR = Q3 - Q1
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR
df = df[(df['chol'] >= lower_bound) & (df['chol'] <= upper_bound)]

In [413]:
# Let's view the box plot after removing outliers of chol column.
# Let's view the cholestrol using interactive boxplot to find outliers
# Create a box plot
fig = px.box(df, y='chol', title='Box Plot of Cholestrol level (chol) of patients after removing Outliers',
             labels={'chol': 'Cholestrol Level (chol)'})

# Show the plot
fig.show()

In [414]:
# Let's jump into the chol column.
print("The average cholestrol level of patients is", df['chol'].mean().astype(int))
print("The maximum cholestrol level of a patient is", df['chol'].max().astype(int))
print("The minimum cholestrol level  of a patient is", df['chol'].min().astype(int))

The average cholestrol level of patients is 242
The maximum cholestrol level of a patient is 369
The minimum cholestrol level  of a patient is 117


In [421]:
print(df.groupby(["sex","dataset"])["chol"].mean().astype(int))

sex     dataset      
Female  Cleveland        252
        Hungary          238
        VA Long Beach    287
Male    Cleveland        239
        Hungary          246
        VA Long Beach    235
Name: chol, dtype: int64


In [416]:
# ✅ Step 1: Group by 'dataset' and 'sex'
df_grouped = df.groupby(['dataset', 'sex'])['chol'].mean().reset_index()

# Step 2: (Optional) Round or convert
df_grouped['chol'] = df_grouped['chol'].astype(int)

# ✅ Step 3: Plot bar chart
fig = px.bar(df_grouped,
             x='dataset',        # 🟡 X-axis is country
             y='chol',           # 🟢 Y-axis is average cholesterol
             color='sex',        # 🔵 Bars colored by sex
             barmode='group',    # Grouped bars (Male/Female side by side)
             title='Average Cholesterol by Sex in Each Country',
             labels={'chol': 'Avg Cholesterol', 'sex': 'Sex', 'dataset': 'Country'},
             hover_data=['chol'])

fig.show()


In [422]:
# let's again have a look on our dataset.
print(df.head())

    id   age     sex    dataset               cp  trestbps   chol    fbs  \
0  1.0  63.0    Male  Cleveland   typical angina     145.0  233.0   True   
1  2.0  67.0    Male  Cleveland     asymptomatic     160.0  286.0  False   
2  3.0  67.0    Male  Cleveland     asymptomatic     120.0  229.0  False   
3  4.0  37.0    Male  Cleveland      non-anginal     130.0  250.0  False   
4  5.0  41.0  Female  Cleveland  atypical angina     130.0  204.0  False   

          restecg  thalch  exang  oldpeak        slope   ca  \
0  lv hypertrophy   150.0  False      2.3  downsloping  0.0   
1  lv hypertrophy   108.0   True      1.5         flat  3.0   
2  lv hypertrophy   129.0   True      2.6         flat  2.0   
3          normal   187.0  False      3.5  downsloping  0.0   
4  lv hypertrophy   172.0  False      1.4    upsloping  0.0   

                thal  num age_group  
0       fixed defect  0.0     60-64  
1             normal  2.0     65-69  
2  reversable defect  1.0     65-69  
3           

In [426]:
# Let's check on our fbs dataset.
df["fbs"].nunique

2

- People who got the heart disease is starting from 28 years and lasting about 77 years.
- Mostly people on an average got the heart disease at the age of 54 years.
- According to dataset, males got heart disease too much as compared to females with percentage of male(78.91%) and female with (21.09%)
- only  males got heart disease at the age of 77 years. 
- Most of the patients from Cleveland which shows that people who live in cleveland(304) got heart disease more as compared to them and the people from Switzerland(123) got disease rarely
- There are 4 types of chest pain that a heart patient experienced
- typical angina pain on avrage in male and female at the age of (59) and non-anginal chest pain in male (54) and female at (51) ad atypical angina pain in male at the age of (51) and in female at (50) and asymptomatic chest pain on average in male at the age of (56) and in female at the age of (56)
- The average resting blood pressure of patients is 132
- The maximum resting blood pressure of a patient is 200
- The minimum resting blood pressure of a patient is 92
- the highest average of resting blood pressure of patients is from VA Long Beech.
- the lowest average of resitng blood pressure of patients is from Switzerland.
- the average cholestrol level of patients is 242.
- the maximum cholestrol is 369 and the lowest one is 117.
- average cholestrol of males(239) females(252) in cleveland,and male(246) female(238) in hungary and in VA Long Beach in males(235) and in females(287)
- the fbs column is about fasting blood sugar it contain two boolean values true and false.if true mean patient also have diabetes otherwise none.